docs/roslyn_best_practices.md

```diff
@@ -0,0 +1,181 @@
+
+Learn more about Async best practices by reading [Async/Await - Best Practices in Asynchronous Programming](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming) by Stephen Cleary
+
+**Do**
+* Leverage the CancellationToken to interrupt processing when cancellation has been requested.
+
+**Do not**
+* Do not force async code to run synchronously. Avoid calling `.Result` and prefer to use await as needed when working in the code fixer construct.
+
+### 5. When using CodeAction.Create prefer Document over Solution
+Minimize the impact of running a code fixer by using the smallest possible scope. 
+
+When writing your CodeFixer you need to create a CodeAction to resolve the diagnostic. If your code fixer only modifies a single document then your CodeAction should return a document.
+
+**Should**
+* Prefer `Task<Document>` when possible and use `Task<Solution>` as needed when returning a result from a Code Fixer.
+
+### 6. Use SolutionEditor and DocEditor when you need to batch changes
+When you need to make numerous changes you will want to batch those changes with the SolutionEditor, or DocEditor. It can be helpful to think about these objects as you would think about using StringBuilder when concatenating a large number of strings. These objects enable you to create a cumulative list of changes that can reduce GC pressure created by modifying immutable SyntaxTrees.
+
+If you’re not batching changes, then most code changes can be accomplished by working directly with the document’s syntax root and the SyntaxGenerator.
+
+
+### 7. Enable support for Fix All
+Don’t force users to fix instances of a diagnostic one-by-one. A FixAll occurrences code fix means: I have a code fix 'C', that fixes a specific instance of diagnostic 'D' in my source and I want to apply this fix to all instances of 'D' across a broader scope, such as a document or a project or the entire solution.
+
+Your code fixer should override `GetFixAllProvider` to return a non-null instance of [FixAll Provider](https://github.com/dotnet/roslyn/blob/main/docs/analyzers/FixAllProvider.md).
+
+### 8. Treat the Code Fixer as a single unit of work
+Code fixers are applied as a one-step change and their order is not guaranteed. If the code you’re adding requires a namespace import then your code fixer should perform that change if necessary. Do not assume that another code fixer will be responsible for adding the correct namespace.
+
+## Considerations specific to upgrade-assistant
+
+There are many stopping points on the journey from .NET Framework to .NET latest and even more scenarios specific to our customers. Because of this, `upgrade-assistant` does not always assume that the code can be compiled. This section highlights some considerations specific to the "work in progress" state that all projects will pass through as they are upgraded.
+
+### 1. Do not assume the code will compile
```

Comment:
I'd also add consider if it's a change that can be done while still targeting .NET Framework. Those can assume things will compile. Only code that we want to function in that transition point should not.

Ie, migrating HttpContext to method injection should still compile and can be done before moving to .NET Core.

Reply:
Wanted to avoid an enumeration of reasons code would not compile here. There are a few, but this one seemed like one that everyone could get onboard with even with minimal awareness of .NET latest.

Reply:
I believe the line of thought I am going down deserves it's own issue and discussion so I created https://github.com/dotnet/upgrade-assistant/issues/604 to track that. Can punt it to there for now :)